source/ssl_poll_perf.c

```diff
@@ -3101,7 +3218,152 @@ client_thread(void)
     return rv;
 }
 
-static void usage(const char *progname)
+static int
+get_run_mode(const char *mode_str)
+{
+    static const char *modes[] = {
+        "client",    /* RUN_MODE_CLIENT == 1 */
+	"server",    /* RUN_MODE_SERVER == 2 */
+        NULL
+    };
+    const char **mode = modes;
+    const char *pattern;
+    const char *ms, *p;
+    int i;
+
+    i = 0;
+    while ((pattern = *mode) != NULL) {
+        ms = mode_str;
+        p = pattern;
+        while (*ms && tolower(*ms) == *p) {
+            ms++;
+            p++;
+        }
+        /*
+         * partial match on pattern.
+         */
+        if (*ms == '\0')
+            return (i + 1);
+
+        mode++;
+	i++;
+    }
+
+    fprintf(stderr,
+            "Unknown mode %s, mode should be either client or server\n",
+            mode_str);
+
+    return -1;
+}
+
+static void
+parse_url(const char *url_in)
+{
+    char *url;
+    char *u = url;
+    const char *http = "https://";
+    const char *h = http;
+    char *hostserve_sep, hostserve_save;
+    char *host_str, *service;
+    int ok;
+    long port_l;
+
+    if (url_in == NULL)
+        errx(1, "%s url where to connect to is missing", __func__);
+
+    url = strdup(url_in);
+    if (url == NULL)
+        errx(1, "%s no mem", __func__);
+
+     /*
+      * chop off http
+      */
+     u = url;
+     while (*u && *h && *h == tolower(*u)) {
+         u++;
+         h++;
+     }
+
+     if (*u == ':') {
+         free(portstr);
+         portstr = strdup("80");
+         if (portstr == NULL)
+             errx(1, "%s no mem", __func__);
+         h++;
+         while (*u && *h && *h == *u) {
+             u++;
+             h++;
+         }
+     } else {
+         free(portstr);
+         portstr = strdup("443");
+         if (portstr == NULL)
+             errx(1, "%s no mem", __func__);
+     }
+
+     if (*h != '\0')
+         errx(1, "%s url (%s) does not start with 'http[s]://'", __func__, url);
+
+     hostserve_sep = strchr(u, '/');
+     if (hostserve_sep != NULL) {
+         hostserve_save = *hostserve_sep;
+         *hostserve_sep = '\0';
+     }
+
+     host_str = NULL;
```

Comment:
nit: imo it's clearer to initialize them to null than explicitly do that, but I'm fine with this approach too